File: python/vectorize/control_reifier.py
from typing import Optional, Dict, List, Tuple
import llvmlite.ir as ir
from control_dependence import ControlCondition, ConditionAnd, ConditionOr
from vloop import VLoop

class ControlReifier:
    def __init__(self, ctx: ir.Context):
        self.ctx = ctx
        self.reified_values: Dict[Tuple[ControlCondition, VLoop], ir.Value] = {}
        self.inserted_insts: List[ir.Instruction] = []

    def reify(self, c: Optional[ControlCondition], vl: VLoop) -> ir.Value:
        if c is None:
            return ir.Constant(ir.IntType(1), 1)

        key = (c, vl)
        if key in self.reified_values:
            return self.reified_values[key]

        if isinstance(c, ConditionAnd):
            self.reify(c.parent, vl)
            cond = c.cond
            if not c.is_true:
                not_inst = ir.IRBuilder(vl.builder.block).not_(cond)
                self.inserted_insts.append(not_inst)
                vl.add_instruction(not_inst, c.parent)
                cond = not_inst
            reified = vl.create_one_hot_phi(c.parent, cond, ir.Constant(ir.IntType(1), 0), "reified.onehot")
        elif isinstance(c, ConditionOr):
            reified = self.reify(c.conds[0], vl)
            for c2 in c.conds[1:]:
                tmp = ir.IRBuilder(vl.builder.block).or_(reified, self.reify(c2, vl))
                self.inserted_insts.append(tmp)
                vl.add_instruction(tmp, None)
                reified = tmp
        else:
            raise ValueError(f"Unsupported control condition type: {type(c)}")

        self.reified_values[key] = reified

        if isinstance(c, ConditionAnd):
            self.reify(c.complement, vl)

        return reified

    def has_value(self, c: Optional[ControlCondition], vl: VLoop) -> bool:
        return c is None or (c, vl) in self.reified_values

    def get_value(self, c: Optional[ControlCondition], vl: VLoop) -> ir.Value:
        assert self.has_value(c, vl)
        if c is None:
            return ir.Constant(ir.IntType(1), 1)
        return self.reified_values[(c, vl)]

File: go/vectorize/control_reifier.go
package hyper

import (
  "github.com/llir/llvm/ir"
  "github.com/llir/llvm/ir/constant"
  "github.com/llir/llvm/ir/types"
  "github.com/llir/llvm/ir/value"
)

type ControlReifier struct {
  Ctx            *types.Context
  ReifiedValues  map[ControlConditionVLoopPair]value.Value
  InsertedInsts  []*ir.Instruction
}

type ControlConditionVLoopPair struct {
  C  *ControlCondition
  VL *VLoop
}

func NewControlReifier(ctx *types.Context) *ControlReifier {
  return &ControlReifier{
    Ctx:           ctx,
    ReifiedValues: make(map[ControlConditionVLoopPair]value.Value),
    InsertedInsts: []*ir.Instruction{},
  }
}

func (cr *ControlReifier) Reify(c *ControlCondition, vl *VLoop) value.Value {
  if c == nil {
    return constant.NewInt(types.I1, 1)
  }
  
  pair := ControlConditionVLoopPair{c, vl}
  if v, ok := cr.ReifiedValues[pair]; ok {
    return v
  }
  
  var reified value.Value
  
  switch cond := c.(type) {
  case *ConditionAnd:
    cr.Reify(cond.Parent, vl)
    condVal := cond.Cond
    if !cond.IsTrue {
      notInst := ir.NewNot(condVal)
      cr.InsertedInsts = append(cr.InsertedInsts, notInst)
      vl.AddInstruction(notInst, cond.Parent)
      condVal = notInst
    }
    reified = vl.CreateOneHotPhi(cond.Parent, condVal, constant.NewInt(types.I1, 0), "reified.onehot")
  
  case *ConditionOr:
    reified = cr.Reify(cond.Conds[0], vl)
    for _, c2 := range cond.Conds[1:] {
      tmp := ir.NewOr(reified, cr.Reify(c2, vl))
      cr.InsertedInsts = append(cr.InsertedInsts, tmp)
      vl.AddInstruction(tmp, nil)
      reified = tmp
    }
  }
  
  cr.ReifiedValues[pair] = reified
  
  if and, ok := c.(*ConditionAnd); ok {
    cr.Reify(and.Complement, vl)
  }
  
  return reified
}

func (cr *ControlReifier) HasValue(c *ControlCondition, vl *VLoop) bool {
  if c == nil {
    return true
  }
  _, ok := cr.ReifiedValues[ControlConditionVLoopPair{c, vl}]
  return ok
}

func (cr *ControlReifier) GetValue(c *ControlCondition, vl *VLoop) value.Value {
  if !cr.HasValue(c, vl) {
    panic("Value not found")
  }
  if c == nil {
    return constant.NewInt(types.I1, 1)
  }
  return cr.ReifiedValues[ControlConditionVLoopPair{c, vl}]
}